### DISCUSSION TITLE
Migrations... postgres to sqlite

### DISCUSSION THREAD

**Discussion by darkone23** [dani-garcia/vaultwarden#3422] [Q&A]
I am moving from a postgres installation to a sqlite installation...

I ported the database using the ruby tool sequel:

`sequel -C <src-postgres> <target-sqlite>`

The sqlite file doesn't have the right migrations so the system blows up on startup. This makes sense looking at the migrations directory - each directory has slightly different migration histories.

I think I will try to create dummy migrations and see if that works... 
Anyone else tried this?

**Comment by BlackDex** [comment]
I suggest to try if you are able to only generate all the inserts. And remove all the schema stuff from your conversion.

Then also remove the migration table from your export.

After that, start Vaultwarden with no sqlite file at all so that it will create a new empty one.

Stop Vaultwarden and then use that sqlite file and your converted file together.

Some basics are also mentioned here.
https://github.com/dani-garcia/vaultwarden/wiki/Migrating-from-MariaDB-%28MySQL%29-to-SQLite#how-to-migrate-from-mariadb-to-sqlite

**Reply by darkone23** [reply]
thank you @BlackDex - I manually diffed the migration histories - going your route probably would have saved me some time :)

I made a sql script I was able to run against my sqlite DB to get the migration system happy and have vaultwarden startup all properly migrated with my data - hooray!

In the case of postgres -> sqlite these were the missing migrations, I inserted them with the following: 

```sql
insert into __diesel_schema_migrations (version, run_on)
select * from (
  with missing_migrations(version) as (
  values                               
        ('20180114171611'),
        ('20180217205753'),
        ('20180427155151'),
        ('20180508161616'),
        ('20180525232323'),
        ('20180601112529'),
        ('20180711181453'),
        ('20180827172114'),
        ('20180910111213'),
        ('20180919144557'),
        ('20181127152651'),
        ('20190526216651')
  )                          
  select * from missing_migrations
  join                                                                                                     
  (select run_on from __diesel_schema_migrations where version = '20190916150000') migrations
  on (1=1)
) migrations_with_timestamps;
```

I kind of wonder if there should maybe be a schema migration to unify the migration histories...
if the schemas are all compatible that is :)

**Reply by M0nk3yOo** [reply]
Do you have an idea why my master password could be wrong after importing the database?
If I login via mobile device anything is fine and decrypted, so I can see my complete vault.

**Reply by BlackDex** [reply]
Probably because of the binary blob format which might differ.
I actually never tried migrating from PostgreSQL to SQLite, only the other way around.

**Reply by M0nk3yOo** [reply]
I also tried to run the commands via shell, but no success.

@darkone23 maybe you've another idea? I don't think that I'm on a complete wrong track, it's just the master password which is not accepted.

**Reply by BlackDex** [reply]
> thank you @BlackDex - I manually diffed the migration histories - going your route probably would have saved me some time :)
> 
> I made a sql script I was able to run against my sqlite DB to get the migration system happy and have vaultwarden startup all properly migrated with my data - hooray!
> 
> In the case of postgres -> sqlite these were the missing migrations, I inserted them with the following:
> 
> ```sql
> insert into __diesel_schema_migrations (version, run_on)
> select * from (
>   with missing_migrations(version) as (
>   values                               
>         ('20180114171611'),
>         ('20180217205753'),
>         ('20180427155151'),
>         ('20180508161616'),
>         ('20180525232323'),
>         ('20180601112529'),
>         ('20180711181453'),
>         ('20180827172114'),
>         ('20180910111213'),
>         ('20180919144557'),
>         ('20181127152651'),
>         ('20190526216651')
>   )                          
>   select * from missing_migrations
>   join                                                                                                     
>   (select run_on from __diesel_schema_migrations where version = '20190916150000') migrations
>   on (1=1)
> ) migrations_with_timestamps;
> ```
> 
> I kind of wonder if there should maybe be a schema migration to unify the migration histories... if the schemas are all compatible that is :)

Not sure how you migrated, but you should have first updated your old db environment, let all migrations complete.

Then, start your instance using the sqlite configuration and let Vaultwarden create a new empty SQLite database.

Then export the PostgreSQL data excluding schema and excluding the migration table.

Somehow convert that data to sqlite compatible format, and import all the records.

This should do the trick.

The schemas are the same code wise, but not fully in the sense that you can convert is between the different database formats that easy.

**Reply by M0nk3yOo** [reply]
Sorry to ask for one more detail. You said "updated your old db environment, let all migrations complete.", do you mean upgrade to 16.0 and... What do you mean with "let all migrations complete".

Upgrade to 16.0 and do an pg_upgrade? (Which I need to work more deeply, because it's not so easy to upgrade docker from 15.4 to 16 and can't run the pg_upgrade afterwards, because the container stops working because of the different DB-version)

**Reply by BlackDex** [reply]
First, Vaultwarden build by this project uses Pg v15.x as a client, so not sure if v16 as a server could pose an issue.

What i meant was the following, all from a Vaultwarden's perspectief.

1. Update Vaultwarden to the latest version using your current database type (pg) and let if run/finish all migrations.
2. Stop Vaultwarden, backup the database.
3. Configure Vaultwarden to use the wanted database type (sqlite) and start Vaultwarden to let it create the database and schema, and directly stop it again.
4. Export all records from the old database (not the schema) excluding the migrations table.
5. Convert this to the target database type and import it into the target database.
6. Start Vaultwarden again, which should now use the new database with the imported data.

**Reply by M0nk3yOo** [reply]
I got it!
Many thanks for your help! The trick was indeed the "blob"-Format(?). After exporting the DB with:

I need to replace another string, too:

```
pg_dump --inserts --data-only -U vaultwardenuser vaultwarden > /var/lib/postgresql/data/db-$(date +%d-%m-%y).backup
sed -i "s/'\\\\x/x'/g" /var/lib/postgresql/data/db-$(date +%d-%m-%y).backup
```

After that I removed any "SET" in the first block, replaced "true" with "'t'" (both without double quotes), and set an "BEGIN;" and "END";
Last I replaced any "public." with "nothing" (NULL/REMOVE/EMPTY)
After that I imported the dump to the sqlite DB, created by vaultwarden


I hope this could help anyone else :-)

**Reply by M0nk3yOo** [reply]
I need to correct one point:

The organization-passwords wasn't visible. The reason for that was to NOT set "true" to "'t'", correct way is to replace that with number "1" (without quotes).

**Reply by bfren** [reply]
+1 to replacing 'public.' - my whole process (with docker compose) was:

1. Set `DATABASE_URL` environment variable to a SQLite file, **bring down the whole stack** then start only Vaultwarden so it creates a fresh `db.sqlite3` with the correct schema, then stop it again.  (My SQLite file is `./v/server/db.sqlite3`)

2. Export from PostgreSQL:
```bash
pg_dump --inserts --data-only -U vaultwardenuser vaultwarden > pg_export.sql
```

3. Strip the migrations table, fix PG hex blob format, and remove the `public.` schema prefix:
```bash
grep -v "__diesel_schema_migrations" pg_export.sql \
  | grep "^INSERT INTO" \
  | sed "s/'\\\\x/x'/g" \
  | sed 's/public\.\([a-zA-Z_]*\)/\1/g' \
  > pg-to-sqlite.sql
```

4. Import into SQLite:
```bash
sqlite3 ./v/server/db.sqlite3 < pg-to-sqlite.sql
```

5. Start Vaultwarden and verify.
